packages/core/FuzzedDataProvider.ts

```diff
@@ -0,0 +1,604 @@
+			range = max / 2.0 - min / 2.0;
+			if (this.consumeBoolean()) {
+				result += range;
+			}
+		} else {
+			range = max - min;
+		}
+		return result + range * this.consumeProbabilityDouble();
+	}
+
+	/**
+	 * Consumes 4 bytes from the fuzzer input.
+	 * @returns a number in the range [0.0, 1.0]
+	 */
+	consumeProbabilityFloat(): number {
+		return this.consumeIntegral(4) / 0xffffffff;
+	}
+
+	/**
+	 * Consumes 8 bytes from the fuzzer input and converts them to an IEEE-754`number`
+	 * in the range [0.0, 1.0].
+	 * @returns a number in the range [0.0, 1.0]
+	 */
+	consumeProbabilityDouble(): number {
+		const n = this.consumeBigIntegral(8, false);
+		const d = (BigInt(0xffffffff) << BigInt(32)) | BigInt(0xffffffff);
+		return Number(n) / Number(d);
+	}
+
+	/**
+	 * Picks an element from `array` based on the fuzzer input.
+	 * Note:The distribution of picks is not perfectly uniform.
+	 * Note: For array sizes > 48 bits, this function will throw an error.
+	 * @param array an `array` of type T to pick an element from.
+	 * @returns an element from `array` chosen based on the fuzzer input
+	 */
+	pickValue<Type>(array: Array<Type>): Type {
+		if (array.length == 0) throw new Error("provided array is empty");
+		return array[this.consumeIntegralInRange(0, array.length - 1)];
+	}
+
+	/**
+	 * -----------------------------------------------------
+	 * Functions for reading data from the front of fuzzer input in
+	 * big-endian order.
+	 * -----------------------------------------------------
+	 */
+
+	/**
+	 * Consumes an array of booleans from the fuzzer input.
+	 * The array might be shorter than requested `maxLength` if the fuzzer input
+	 * is not sufficiently long.
+	 * @param maxLength - the maximum length of the array
+	 * @returns an array of booleans
+	 */
+	consumeBooleans(maxLength: number): boolean[] {
+		const arrayLength = Math.min(this._remainingBytes, maxLength);
+		const result = new Array<boolean>(arrayLength);
+		for (let i = 0; i < arrayLength; i++) {
+			result[i] = (this.data[this.dataPtr + i] & 1) == 1;
+		}
+		this._remainingBytes -= arrayLength;
+		this.dataPtr += arrayLength;
+		return result;
+	}
+
+	/**
+	 * Consumes an array of integrals from fuzzer data.
+	 * The array might be shorter than requested `maxLength` if the fuzzer input
+	 * is not sufficiently long.
+	 * @param maxLength - number of integers to consume
+	 * @param numBytesPerIntegral - number of bytes to consume for each integral
+	 * @param isSigned - whether the integrals are signed
+	 * @returns an array of integrals
+	 */
+	consumeIntegrals(
+		maxLength: number,
+		numBytesPerIntegral: number,
+		isSigned = false
+	): number[] {
+		const availableBytes = Math.min(
+			this._remainingBytes,
+			maxLength * numBytesPerIntegral
+		);
+		const arrayLength = Math.ceil(availableBytes / numBytesPerIntegral);
+		const result = new Array<number>(arrayLength);
+		for (let i = 0; i < arrayLength; i++) {
+			result[i] = this.consumeIntegralLEorBE(
```

Comment:
The method structure seems quite similar in most of the methods. Calculate `availableBytes`, `arrayLength`, execute another method to read a value and store it in a result array. 
Are the methods similar enough to extract this logic out into a dedicated higher-order function?

Reply:
Added a helper function for that in 26b3056.

Reply:
Ah, sorry, that was misleading. I was thinking about something like `consume(maxLength, bytesPerNumber, isSigned, consumeIntegralLEorBE) `, so extract out the loop logic. But perhaps the functions are too different for that.

Reply:
Added 6f73f5f, what do you think?
I think we should benchmark the performance at some point, to see whether JIT compiler can efficiently optimize all these layers of function calls away. Since methods in FuzzedDataProvider will be called trillions of times (I'm projecting into the future a bit :), it might make sense to optimize for speed if possible.

Reply:
Hmm, well, that doesn't change so much and, as you mentioned, could lead to performance problems. We should probably leave it as it was, shouldn't we?